<<<<<<< HEAD
# workingMetaFinisherSC
=======
# MetaFinisherSC
To run the tool, you should have input files(contigs.fasta, raw_reads.fasta) in destinedFolder. The tool will first run through FinisherSC to generate improved3.fasta to destinedFolder. Then it will go through the newly added functions suitable for metagenomics study and the final output file is **abun.fasta**
The command to run MetaFinisherSC is 

        python metaFinisherSC.py destinedFolder/ mummerPath/
        
A test set is located here https://www.dropbox.com/sh/slahcgv55037aiv/AAAzkaP6HILdBlH2G_xapKYHa?dl=0  

        
>>>>>>> FETCH_HEAD
